File: app/src/main/java/com/example/khughes/machewidget/APIMPSService.java
package com.example.khughes.machewidget;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.Headers;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Query;

public interface APIMPSService {

    @Headers({"Content-Type: application/json",
            "Accept-Language: en-US", "Application-Id: " + Constants.APID,
            "Authorization: Basic ZWFpLWNsaWVudDo="})
    @POST("token/v2/cat-with-ci-access-token")
    Call<AccessToken> getAccessToken(@Body RequestBody token);

    @Headers({"Content-Type: application/json",
            "Accept-Language: en-US", "Application-Id: " + Constants.APID,
            "Authorization: Basic ZWFpLWNsaWVudDo="})
    @POST("token/v2/cat-with-refresh-token")
    Call<AccessToken> refreshAccessToken(@Body RequestBody token);

    @Headers({"Content-Type: application/json",
            "Accept-Language: en-US", "Application-Id: " + Constants.APID,
            "Authorization: Basic ZWFpLWNsaWVudDo="})
    @GET("users")
    Call<AccessToken> getUserProfile(@Header("auth-token") String token);

    @Headers({
            "locale: en-US",
            "user-agent: FordPass/5 CFNetwork/1327.0.4 Chrome/96.0.4664.110",
    })
    @POST("expdashboard/v1/details")
    Call<UserDetails> getUserDetails(@Header("auth-token") String token,
                                      @Header("application-id") String APID,
                                      @Header("countrycode") String country,
                                      @Body RequestBody data);
}

File: app/src/main/java/com/example/khughes/machewidget/APIMPSService.kt
package com.example.khughes.machewidget

import com.google.gson.annotations.Expose
import javax.annotation.Generated

import okhttp3.RequestBody
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.http.*

@Generated("jsonschema2pojo")
data class VehicleList (
    val status: String,
    @Expose
    val vehicles: List<VehicleData>,
)

@Generated("jsonschema2pojo")
data class VehicleData(
    val vehicleId: String,
    val make: String,
    val modelName: String,
    val modelYear: String,
    val color: String,
    val nickName: String,
    val modemEnabled: Boolean,
    val vehicleAuthorizationIndicator: Long,
    val serviceCompatible: Boolean,
    val engineType: String,
)

interface APIMPSService {

    @Headers(
        "Accept: application/json",
        "Content-type: application/json",
        "Application-id: " + FordConnectConstants.APID
    )
    @GET("fordconnect/v3/vehicles")
    fun getVehicleList (@Header("Authorization") accessToken: String) : Call<VehicleList>?

    @Headers(
        "Application-id: " + FordConnectConstants.APID
    )
    @GET("fordconnect/v3/vehicles/{vehicleID}")
    fun getStatus (
        @Path("vehicleID") vehicleId: String,
        @Header("Authorization") accessToken: String
    ) : Call<CarStatus>?

    @Headers(
        "Application-id: " + FordConnectConstants.APID
    )
    @GET("fordconnect/v3/vehicles/{vehicleID}/vin")
    fun getVIN (
        @Path("vehicleID") vehicleId: String,
        @Header("Authorization") accessToken: String
    ) : Call<ResponseBody>?

    @Headers(
        "Application-id: " + FordConnectConstants.APID
    )
    @GET("fordconnect/v1/vehicles/{vehicleID}/images/full")
    fun getVehicleImage(
        @Path("vehicleID") vehicleId: String,
        @Header("Authorization") accessToken: String,
        @Query("make") make: String?,
        @Query("model") model: String?,
        @Query("year") year: String?,
    ): Call<ResponseBody?>?


}